include/deal.II/matrix_free/constraint_info.h

```diff
@@ -0,0 +1,478 @@
+        };
+
+        // plain indices
+        plain_dof_indices_per_cell[cell_no] = local_dof_indices_lex;
+        for (auto &i : plain_dof_indices_per_cell[cell_no])
+          i = global_to_local(i);
+
+        if (hanging_nodes)
+          {
+            AssertIndexRange(cell_no,
+                             this->hanging_node_constraint_masks.size());
+            AssertIndexRange(cell_no, this->active_fe_indices.size());
+
+            std::vector<internal::MatrixFreeFunctions::ConstraintKinds> mask(
+              cell->get_fe().n_components());
+            hanging_nodes->setup_constraints(
+              cell, {}, lexicographic_numbering, local_dof_indices_lex, mask);
+
+            hanging_node_constraint_masks[cell_no] = mask[0];
+            active_fe_indices[cell_no]             = cell->active_fe_index();
+          }
+
+        for (unsigned int i = 0; i < local_dof_indices_lex.size(); ++i)
+          {
+            types::global_dof_index current_dof = local_dof_indices_lex[i];
+            const auto *            entries_ptr =
+              constraints.get_constraint_entries(current_dof);
+
+            // dof is constrained
+            if (entries_ptr != nullptr)
+              {
+                const auto &                  entries   = *entries_ptr;
+                const types::global_dof_index n_entries = entries.size();
+                if (n_entries == 1 &&
+                    std::abs(entries[0].second - 1.) <
+                      100 * std::numeric_limits<double>::epsilon())
+                  {
+                    current_dof = entries[0].first;
+                    goto no_constraint;
+                  }
+
+                constraint_indicator.push_back(constraint_iterator);
+                constraint_indicator.back().second =
+                  constraint_values.insert_entries(entries);
+
+                // reset constraint iterator for next round
+                constraint_iterator.first = 0;
+
+                if (n_entries > 0)
+                  {
+                    const std::vector<types::global_dof_index> &
+                      constraint_indices = constraint_values.constraint_indices;
+                    for (unsigned int j = 0; j < n_entries; ++j)
+                      {
+                        dof_indices.push_back(
+                          global_to_local(constraint_indices[j]));
+                      }
+                  }
+              }
+            else
+              {
+              no_constraint:
+                dof_indices.push_back(global_to_local(current_dof));
+
+                // make sure constraint_iterator.first is always within the
+                // bounds of unsigned short
+                Assert(constraint_iterator.first <
+                         (1 << (8 * sizeof(unsigned short))) - 1,
+                       ExcInternalError());
+                constraint_iterator.first++;
+              }
+          }
+      }
+
+      void
+      finalize()
```

Comment:
A difference to the setup in the original code is that I am first collecting everything in vectors of vectors and after that convert them to CRS data structures. I do this to put the data to the right position during the cell loop in the setup of the global-coarsening transfer. But what could do is to loop over all cells category by category or  alternatively to sort the data in a second step. Both needs some more involved refactoring in `MGTwoLevelTransfer` so that I would like to postpone this to a follow-up PR!